clash-lib/src/Clash/Driver/Manifest.hs

```diff
@@ -0,0 +1,274 @@
+
+mkManifest ::
+  -- | Options Clash was run with
+  ClashOpts ->
+  -- | Component of top entity
+  Component ->
+  -- | All other entities
+  [Component] ->
+  -- | Files and  their hashes
+  [(FilePath, ByteString)] ->
+  -- | Hash returned by 'readFreshManifest'
+  Int ->
+  -- | New manifest
+  Manifest
+mkManifest ClashOpts{..} Component{..} components files topHash = Manifest
+  { manifestHash = topHash
+  , portInNames = map Id.toText topInNames
+  , portOutNames = map Id.toText topOutNames
+  , componentNames = map Id.toText compNames
+  , topComponent = Id.toText componentName
+  , fileNames = files
+  , successFlags = (opt_inlineLimit, opt_specLimit, opt_floatSupport)
+  }
+ where
+  topInNames = map fst inputs
+  topOutNames = map (fst . (\(_,x,_) -> x)) outputs
+  compNames = map Netlist.componentName components
+
+-- | Pretty print an unexpected modification as a list item.
+pprintUnexpectedModification :: UnexpectedModification -> String
+pprintUnexpectedModification = \case
+  Modified p -> "Unexpected modification in " <> p
+  Added p -> "Unexpected extra file " <> p
+  Removed p -> "Unexpected removed file " <> p
+
+-- | Pretty print a list of unexpected modifications. Print a maximum of /n/
+-- modifications.
+pprintUnexpectedModifications :: Int -> [UnexpectedModification] -> String
+pprintUnexpectedModifications 0 us = pprintUnexpectedModifications maxBound us
+pprintUnexpectedModifications _ [] = []
+pprintUnexpectedModifications _ [u] = "* " <> pprintUnexpectedModification u
+pprintUnexpectedModifications 1 (u:us) =
+  "* and " <> show (length (u:us)) <> " more unexpected changes"
+pprintUnexpectedModifications n (u:us) =
+  "* " <> pprintUnexpectedModification u
+        <> "\n" <> pprintUnexpectedModifications (n-1) us
+
+-- | Reads a manifest file. Does not return manifest file if:
+--
+--  * Caching is disabled through @-fclash-no-cache@.
+--  * Manifest could not be found.
+--  * Cache is stale. This could be triggered by any of the given arguments.
+--
+-- Raises an exception if the manifest file or any of the files it is referring
+-- to was inaccessible.
+--
+readFreshManifest ::
+  -- | "This" top entity plus all that depend on it.
+  [TopEntityT] ->
+  -- | Core expressions and entry point. Any changes in the call graph will
+  -- trigger a recompile.
+  (BindingMap, Id) ->
+  -- | Any changes in any primitive will trigger a recompile.
+  CompiledPrimMap ->
+  -- | Certain options will trigger recompiles if changed
+  ClashOpts ->
+  -- | Clash modification date
+  UTCTime ->
+  -- | Path to manifest file.
+  FilePath ->
+  -- | ( Nothing if no manifest file was found
+  --   , Nothing on stale cache, disabled cache, or not manifest file found )
+  IO (Maybe [UnexpectedModification], Maybe Manifest, Int)
+readFreshManifest tops (bindingsMap, topId) primMap opts@(ClashOpts{..}) clashModDate path = do
+  manifestM <- readManifest path
+  modificationsM <- sequence (isUserModified path <$> manifestM)
```

Comment:
`traverse (isUserModified path) manifestM`

Reply:
I always forget `traverse` exists

Reply:
It's just `mapM`! But with `Applicative` instead of `Monad` as the constraint